sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/implementation/util/ShareServiceVersionPolicy.java

```diff
@@ -0,0 +1,75 @@
+package com.azure.storage.file.share.implementation.util;
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.storage.common.implementation.Constants;
+import com.azure.storage.file.share.ShareServiceVersion;
+import reactor.core.publisher.Mono;
+
+import java.net.URL;
+
+import static com.azure.storage.common.implementation.StorageImplUtils.throwIfContainsHeader;
+
+public class ShareServiceVersionPolicy implements HttpPipelinePolicy {
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        HttpHeaders requestHeaders = context.getHttpRequest().getHeaders();
+        URL requestUrl = context.getHttpRequest().getUrl();
+        ShareServiceVersion serviceVersion = toServiceVersion(
+            requestHeaders.getValue(Constants.HeaderConstants.SERVICE_VERSION));
+        String ANY_FILE_API = "any file API";
+
+        if (serviceVersion.ordinal() == ShareServiceVersion.getLatest().ordinal()) {
+            return next.process();
+        }
+
+        // 2019_07_07 check
+        if (serviceVersion.ordinal() < ShareServiceVersion.V2019_07_07.ordinal()) {
+            // File lease
+            throwIfContainsHeader(requestHeaders, "x-ms-lease-id", ANY_FILE_API, serviceVersion.getVersion());
+            throwIfContainsHeader(requestHeaders, "x-ms-lease-duration", ANY_FILE_API, serviceVersion.getVersion());
+            throwIfContainsHeader(requestHeaders, "x-ms-proposed-lease-id", ANY_FILE_API, serviceVersion.getVersion());
+
+            if (requestUrl.getQuery() != null && requestUrl.getQuery().contains("comp=lease")) {
```

Comment:
Should a helper method be added into `StorageImplUtils` that validates that the request doesn't contain a given query parameter?

Reply:
yeah that's a good idea.